core/src/software/aws/toolkits/core/credentials/sso/DiskCache.kt

```diff
@@ -0,0 +1,117 @@
+        customDateModule.addDeserializer(Instant::class.java,
+            CliCompatibleInstantDeserializer()
+        )
+
+        it.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
+        it.registerModule(JavaTimeModule())
+        it.registerModule(customDateModule) // Override the Instant deserializer with custom one
+        it.dateFormat = StdDateFormat().withTimeZone(TimeZone.getTimeZone(ZoneOffset.UTC))
+    }
+
+    override fun loadClientRegistration(ssoRegion: String): ClientRegistration? {
+        val cacheFile = cacheDir.resolve(clientRegistrationCacheKey(ssoRegion))
+        val inputStream = cacheFile.inputStreamIfExists() ?: return null
+        return tryOrNull {
+            val clientRegistration = objectMapper.readValue<ClientRegistration>(inputStream)
+            if (clientRegistration.expiresAt.isAfter(Instant.now(clock))) {
+                clientRegistration
+            } else {
+                null
+            }
+        }
+    }
+
+    override fun saveClientRegistration(ssoRegion: String, registration: ClientRegistration) {
+        cacheDir.resolve(clientRegistrationCacheKey(ssoRegion)).outputStream().use {
+            objectMapper.writeValue(it, registration)
+        }
+    }
+
+    override fun invalidateClientRegistration(ssoRegion: String) {
+        cacheDir.resolve(clientRegistrationCacheKey(ssoRegion)).deleteIfExists()
+    }
+
+    private fun clientRegistrationCacheKey(ssoRegion: String): String = "aws-toolkit-jetbrains-$ssoRegion.json"
+
+    override fun loadAccessToken(ssoUrl: String): AccessToken? {
+        val cacheFile = cacheDir.resolve(accessKeyCacheKey(ssoUrl))
+        val inputStream = cacheFile.inputStreamIfExists() ?: return null
+
+        return tryOrNull {
+            val clientRegistration = objectMapper.readValue<AccessToken>(inputStream)
+            if (clientRegistration.expiresAt.isAfter(Instant.now(clock))) {
+                clientRegistration
+            } else {
+                null
+            }
+        }
+    }
+
+    override fun saveAccessToken(ssoUrl: String, accessToken: AccessToken) {
+        cacheDir.resolve(accessKeyCacheKey(ssoUrl)).outputStream().use {
+            objectMapper.writeValue(it, accessToken)
+        }
+    }
+
+    override fun invalidateAccessToken(ssoUrl: String) {
+        cacheDir.resolve(accessKeyCacheKey(ssoUrl)).deleteIfExists()
+    }
+
+    private fun accessKeyCacheKey(ssoUrl: String): String {
+        val digest = MessageDigest.getInstance("SHA-1")
+        val sha = digest.digest(ssoUrl.toByteArray(Charsets.UTF_8)).toHexString()
+        return "$sha.json"
+    }
+
+    private class CliCompatibleInstantDeserializer : StdDeserializer<Instant>(Instant::class.java) {
+        override fun deserialize(parser: JsonParser, context: DeserializationContext): Instant {
+            val dateString = parser.valueAsString
+
+            // CLI appends UTC, which Java refuses to parse. Convert it to a Z
```

Comment:
do we know why CLI does this instead of standard ISO-8601? can the CLI parse the 'Z'? maybe we should write UTC instead of Z too?

Reply:
Yes, they can parse the Z. Their lib is more lenient